src/wallet/spend.cpp

```diff
@@ -1260,6 +1263,7 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
     // accidental reuse.
     reservedest.KeepDestination();
 
+    wallet.WalletLogPrintf("Coin Selection: Algorithm:%s, Waste:%d\n", GetAlgorithmName(result.GetAlgo()), result.GetWaste());
```

Comment:
My concern is less with the specific wording and more that we are logging low-level algorithm internals that at best are not actionable for the node runner and at worst confusing.

For developers, the algorithm and waste metric (along with additional data points) are already captured in `TRACE5`:

https://github.com/bitcoin/bitcoin/blob/efb54a2bd284639722da794650d2afda66fb0770/src/wallet/spend.cpp#L1128

So my questions are:

1. How is the introduction of this logline related to this _specific_ SFFO and BnB bugfix
2. How does it help in a way that the already existing tracepoint couldn't have helped?

In general, I think we should have a strong justification for adding logs and I don't see that here.

Reply:
Tracepoints are OS level hooks that are not part of the software by default. Logging is about keeping the history of events that occurred along the software lifetime. This does not only involve errors, it also involves information about the soft activity.
Unlike logging, which can be enabled on-the-fly by simply calling an RPC command, users cannot enable tracepoints on their node without recompiling the software with a specific flag. Which requires compiling knowledge, forces the user to shutdown the node, etc..

Usually, when users encounter an issue, we require further information to assess them properly. Thus we ask for the logs. This events record need to provide us sufficient information to understand the behavior of the system so we can provide better assessment to our users, and also to encounter issues remotely.
In the context of this PR, coin selection is not deterministic and its behavior is not recorded anywhere. Thus, we have no way of knowing how the process operated, making it more challenging to understand and find bugs than it should be.

By logging the coin selection algorithm name and the waste score (at least), we gain information that we currently don't have. This information can be useful for identifying issues and explaining behaviors that may occur only once and be hard to replicate.

> My concern is less with the specific wording and more that we are logging low-level algorithm internals that at best are not actionable for the node runner and at worst confusing.

Not all logs need to be actionable by the node runner. In fact, most of them are not. The messages that are meant to provide an actionable path for the user are logged by default, the rest are under specific logging domain. Existing to help us detect issues and describe behaviors remotely.
What I mean with this is that node runners need to explicitly enable the wallet logging to see the introduced log. Which, usually, they only do upon request when they are facing a problem.

Reply:
The logs are useful for figuring out what happened when a user asks about why their wallet chose a crazy number of UTXOs when one would have been sufficient. As we add more algos, being able to see in the logs what algo was used, and what the resulting waste score was, is helpful for investigating such questions.

Reply:
> wallet chose a crazy number of UTXOs when one would have been sufficient

I don't see how this example is relevant to this PR, which is why I think the logging discussion would be better moved to a logging PR.

This also presumes the user already had the logging enabled before the issue happened. I'm not sure the costs outweigh the benefits of having this log always on to be able to answer questions like your example.

Reply:
Again, not convinced having this log on all the time is worth it, and don't see how this log is relevant to this PR.

Reply:
> My point is that for developers, we already have tracepoints as a tool for debugging and observability into low-level details of how the software is working. In many cases, a developer will try to reproduce an issue reported by a user on their own node, at which point debuggers and tracepoints are going to be more useful than logs.

Replicate issues without information is by far more challenging than replicating them with information. It is a blind guess.

Also, what you are saying works (up to a certain point) for deterministic procedures but not for non-deterministic ones.
Coin selection, like many other processes, isn't deterministic, and usually, we cannot replicate exactly what happened to the user.
Moreover, you are not taking into account the sensitivity of wallet information. There is data we can request in public, and some other data that is safer not to request, affecting how wallet behaviors can be understood remotely.

> Again, not convinced having this log on all the time is worth it, and don't see how this log is relevant to this PR.

Back to the original bug fix PR #28395, this issue provokes the transaction creation process to generate a change output for a set of BnB selected coins, which is unexpected and incorrect. BnB is specialized for changeless solutions.

In case this issue would had arose in one of our user's wallet, logging this information could have helped us to detect the problem remotely. We would have noticed that the transaction creation process produced a change output after selecting coins using BnB. And start our investigation from there. Eliminating any blind guess start.

Reply:
Looking at #28395, the bug is related to SFFO. If we are going to log something, it would be better to log information related to what options the user selected, such as logging that SFFO was selected (I did a quick grep and didn't see that this was being logged, but might have missed it). In general, I think it's always better to log the parameters that were used to start a process, rather than log the resulting algorithm internals.

If we expect BnB to _always_ produce changeless solutions, why aren't we instead checking this assumption after coin selection and logging an error, or at least a warning, that BnB produced change? That would make more sense to me than what you currently have.

Just to reiterate in case it's not clear: I'm not opposed to logging. I just don't think adding this specific log here helps at all and I think if we are going to add logs, let's take a more holistic look at what's being logged or not logged in the wallet, what's is most actionable for troubleshooting, and decide what levels things should be logged at in separate PR.

Reply:
I think @josibake is making a good point here. Algorithm and waste score by itself would probably still make it difficult to learn something about what went wrong (although they would be useful if the issue filer also provided the corresponding transaction). It might be better to also log the parameters for the coin selection attempt:

- feerate
- sffo (true/false)
- preset inputs (true/false)
- resulting input count (per type?)
- change created (true/false)
- algo that lead to solution
- waste score

Don’t log: exact utxos spent or created, payment amount, change amount, count of available UTXOs, wallet balance, etc.

Reply:
> it would be better to log information related to what options the user selected, such as logging that SFFO was selected (I did a quick grep and didn't see that this was being logged, but might have missed it). In general, I think it's always better to log the parameters that were used to start a process, rather than log the resulting algorithm internals.

That's not correct. 
The wallet stores its own state linked to the inner node's state. Logging users' RPC command inputs only provides limited information. These inputs can easily be requested on-demand. Anyone can copy-paste the RPC command line in a public place (hiding the sensitive data).
What users cannot do is provide all wallet internal information publicly, which is used throughout any wallet process. Furthermore, and once more, the coin selection procedure is not deterministic. Logging only the high-level user inputs wouldn't help with non-deterministic incorrect behaviors, failures, or crashes.

> If we expect BnB to always produce changeless solutions, why aren't we instead checking this assumption after coin selection and logging an error, or at least a warning, that BnB produced change? That would make more sense to me than what you currently have.

What you are suggesting is ok for a post-mortem comment. You mindset is set as if we would be knowing all things that could happen, and thats incorrect. I wish we would have a crystal ball to see the future and include the same checks for all possible issues. The truth is that code is written by people, and AIs, and people make mistakes and forget about including checks. And logging certain specific internal events help troubleshooting issues.
In this case, the logging would have helped us.

> let's take a more holistic look at what's being logged or not logged in the wallet, what's is most actionable for troubleshooting, and decide what levels things should be logged at in separate PR.

Thats a separate topic. Wouldn't be bad to have someone doing it. Go ahead with it.

Reply:
> What you are suggesting is ok for a post-mortem comment.

I think you're misunderstanding my point. I am saying in the context of this PR (which is a bug fix), I don't think the log line you are adding here adds value and that a better thing to add in this PR would be a check that BnB is never returning change, which either errors or logs a warning.

Reply:
Right, but this is my concern: a bug is found and fixed, and then a log is added that _might_ have helped detect the already fixed bug (in this case, I'm still not convinced this log as written would have helped much). This pattern tends toward log bloat over time.

What I have been suggesting from the beginning is that we keep this PR focused on the bug fix and address logging in a separate PR where we can take a more holistic look. I do think logging the input parameters for coin selection is going to be far more helpful in troubleshooting these types of issues in the future, and also think instead of logs we should prefer adding assumption checks that either fail quickly or warn the user.

At this point, I think we just disagree on the usefulness of logs. I'd still prefer if we left this commit off this PR, but I'm not going to argue the point any further.

Reply:
> Right, but this is my concern: a bug is found and fixed, and then a log is added that might have helped detect the already fixed bug (in this case, I'm still not convinced this log as written would have helped much). This pattern tends toward log bloat over time.

Well, thats not what occurs in practice. Isn't a pattern at all. Could check all solved bugs in the wallet the past ~2 years (since I'm around at least) and see how many log lines were added to the wallet. I'm quite confident that we did not introduce non-useful logging lines there.

> I do think logging the input parameters for coin selection is going to be far more helpful in troubleshooting these types of issues in the future

Glad you are on-board about logging internal processes information now :).

>  instead of logs we should prefer adding assumption checks that either fail quickly or warn the user.

Thats for sure, hard to disagree. But all of us make mistakes writing code; checks could be forgotten. And processes could be none deterministic. Logging certain specific information helps troubleshooting this situations.

> At this point, I think we just disagree on the usefulness of logs. I'd still prefer if we left this commit off this PR, but I'm not going to argue the point any further.

Sure. I'm actually stronger for it now.